src/Adapter/DoctrineItems.php

```diff
@@ -0,0 +1,354 @@
+                }
+
+                $next = $current->next();
+
+                if ($item->getVersion() > $next) {
+                    throw new UnexpectedVersionNumber($item->getId(), $item->getVersion(), $next);
+                }
+
+                if ($item->getVersion() < $next) {
+                    $this->updateVersion($item);
+
+                    return;
+                }
+
+                $this->insertVersion($item);
+            }
+        );
+    }
+
+    public function remove(ItemId $id, ?ItemVersionNumber $version) : void
+    {
+        $this->connection->transactional(
+            function () use ($id, $version) : void {
+                if (null === $version) {
+                    $this->deleteItem($id);
+
+                    return;
+                }
+
+                try {
+                    $current = $this->getCurrentVersion($id, LockMode::PESSIMISTIC_WRITE);
+                } catch (ItemNotFound $e) {
+                    return;
+                }
+
+                if ($version > $current) {
+                    return;
+                }
+
+                if ($version < $current) {
+                    throw new UnexpectedVersionNumber($id, $version, $current);
+                }
+
+                if (1 === $current->toInt()) {
+                    $this->deleteItem($id);
+
+                    return;
+                }
+
+                $this->deleteVersion($id, $current);
+            }
+        );
+    }
+
+    public function get(ItemId $id, ?ItemVersionNumber $version = null) : ItemVersion
+    {
+        if (null === $version) {
+            $sql = 'SELECT version.id, version.version, version.content, version.hash'
+                ." FROM {$this->tableItems} item LEFT JOIN {$this->tableVersions} version"
+                .' ON item.id = version.id AND item.version = version.version'
+                .' WHERE item.id = :id';
+            $stmt = $this->connection->executeQuery($sql, ['id' => (string) $id]);
+
+            $row = $stmt->fetch(FetchMode::ASSOCIATIVE);
+
+            if (false === $row) {
+                throw new ItemNotFound($id);
+            }
+
+            return $this->rowToItemVersion($row);
+        }
+
+        $sql = 'SELECT item.id, version.version, version.content, version.hash'
+            ." FROM {$this->tableItems} item LEFT JOIN {$this->tableVersions} version"
+            .' ON item.id = version.id AND version.version = :version WHERE item.id = :id';
+        $stmt = $this->connection->executeQuery($sql, ['id' => (string) $id, 'version' => $version->toInt()]);
+
+        $row = $stmt->fetch(FetchMode::ASSOCIATIVE);
+
+        if (!$row) {
+            throw new ItemNotFound($id);
+        }
+
+        if (null === $row['version']) {
+            throw new VersionNotFound($id, $version);
+        }
+
+        return $this->rowToItemVersion($row);
+    }
+
+    public function list(int $limit = 10, ?string $cursor = null) : ItemListPage
```

Comment:
I notice there is no `asc`|`desc` exposed to the interface, but would specify the one chosen (`desc`?) in `$query` to avoid people having to look it up elsewhere

Reply:
Unordered in the business sense, the SQL order here is first version insertion order https://github.com/libero/content-api-bundle/blob/461140aea8ea37c9241986d306c2ef21e41bdac3/src/Adapter/DoctrineItems.php#L163

(Filesystem adapter is ID alphabetical. So both deterministic.)

Reply:
Now seen it, that's deterministic yes.